dev/benchmarks/test_apps/stocks/lib/main.dart

```diff
@@ -65,49 +45,161 @@ class StocksAppState extends State<StocksApp> {
-      routes: <String, WidgetBuilder>{
-         '/':         (BuildContext context) => StockHome(stocks, _configuration, configurationUpdater),
-         '/settings': (BuildContext context) => StockSettings(_configuration, configurationUpdater),
-      },
-      onGenerateRoute: _getRoute,
+      debugShowMaterialGrid: configuration.debugShowGrid,
+      showPerformanceOverlay: configuration.showPerformanceOverlay,
+      showSemanticsDebugger: configuration.showSemanticsDebugger,
+      routeNameParser: StockRouteNameParser(),
+      routerDelegate: StockRouterDelegate(_state),
+    );
+  }
+}
+
+class StockRouteNameParser extends RouteNameParser<Uri> {
+  @override
+  Future<Uri> parse(String routeName) {
+    final List<String> segments = routeName.split('?');
+    final String path = segments[0];
+    final String query = segments.length > 1 ? segments[1] : '';
+    final Uri syntheticUrl = Uri(path: path, query: query);
+    return SynchronousFuture<Uri>(syntheticUrl);
+  }
+
+  @override
+  String restore(Uri configuration) {
+    return configuration.toString();
+  }
+}
+
+class StockRouterDelegate extends RouterDelegate<Uri> with PopNavigatorRouterDelegateMixin<Uri>{
+  StockRouterDelegate(
+    this.stockState
+  ) : navigatorKey = GlobalObjectKey<NavigatorState>(stockState) {
+    stockState.addListener(notifyListeners);
+  }
+  StockState stockState;
+
+  final HeroController _heroController = MaterialApp.createMaterialHeroController();
+
+  @override
+  final GlobalObjectKey<NavigatorState> navigatorKey;
+
+  @override
+  Uri get currentConfiguration {
+    assert(stockState.initialized);
+    Map<String, List<String>> queries;
+    String routeName;
+    if (stockState.stockSymbol != null) {
+      queries = <String, List<String>>{
+        'symbol': <String>[stockState.stockSymbol],
+      };
+      routeName = '/stock';
+    } else {
+      routeName = stockState.inSettings ? '/settings' : '/';
+    }
+    return Uri(
+      path: routeName,
+      queryParameters: queries,
+    );
+  }
+
+  @override
+  Future<void> setNewRoutePath(Uri configuration) {
+    assert(configuration != null);
+    if (configuration.queryParameters != null) {
+      final String symbols = configuration.queryParameters['symbol'];
+      if (symbols != null && configuration.path == '/stock') {
+        stockState.stockSymbol = symbols[0];
+        stockState.inSettings = false;
+        return SynchronousFuture<void>(null);
+      }
+    }
+
+    stockState.stockSymbol = null;
+    stockState.inSettings = configuration.path == '/settings';
+    return SynchronousFuture<void>(null);
+  }
+
+  bool _handlePopPage(Route<dynamic> route, dynamic result) {
+    final bool success = route.didPop(result);
+    if (success) {
+      final Page<void> page = route.settings as Page<void>;
+      if (page is StockSettingsPage)
+        stockState.inSettings = false;
+      if (page is StockPage)
+        stockState.stockSymbol = null;
+    }
+    return success;
```

Comment:
seems like we should be able to drive this entirely from the stockState, no need to look at the route we just popped, right? at best we should assert that the page is what we expect given the state

Reply:
we can if we use appstate to control the pop up too.  See https://github.com/flutter/flutter/blob/e5b93faf2f60189139d1e8a61065b3a62e5cf283/dev/benchmarks/test_apps/stocks/lib/stock_settings.dart#L67

If not, then we still have to wire up the back button dispatcher to navigator.pop. I think this is easier for people and less error prompt. What do you think?

Reply:
this shouldn't be getting called for non-Page routes

Reply:
This won't be called, but in order to support non-page route we have to wire the back button dispatcher to navigator.pop thus we need to provide this method

Reply:
My point is that by the time this specific method gets called, you should know that the page that's top is the one that's described by the app state, because if it wasn't the method wouldn't get called (the navigator would take care of it). And so you can just update the app state rather than examining the page.

Reply:
Ah correct, I misunderstood the question.